src/coreclr/jit/codegenarm.cpp

```diff
@@ -653,6 +658,254 @@ void CodeGen::genJumpTable(GenTree* treeNode)
     genProduceReg(treeNode);
 }
 
+//------------------------------------------------------------------------
+// genLockedInstructions: Generate code for a GT_XADD or GT_XCHG node.
+//
+// Arguments:
+//    treeNode - the GT_XADD/XCHG node
+//
+void CodeGen::genLockedInstructions(GenTreeOp* treeNode)
+{
+    GenTree*  data      = treeNode->AsOp()->gtOp2;
+    GenTree*  addr      = treeNode->AsOp()->gtOp1;
+    regNumber targetReg = treeNode->GetRegNum();
+    regNumber dataReg   = data->GetRegNum();
+    regNumber addrReg   = addr->GetRegNum();
+
+    genConsumeAddress(addr);
+    genConsumeRegs(data);
+
+    assert(!treeNode->OperIs(GT_XORR, GT_XAND));
+    assert(treeNode->OperIs(GT_XCHG) || !varTypeIsSmall(treeNode->TypeGet()));
+
+    emitAttr dataSize = emitActualTypeSize(data);
+
+    regNumber tempReg  = treeNode->ExtractTempReg(RBM_ALLINT);
+    regNumber storeReg = (treeNode->OperGet() == GT_XCHG) ? dataReg : treeNode->ExtractTempReg(RBM_ALLINT);
+    regNumber loadReg  = (targetReg != REG_NA) ? targetReg : storeReg;
+
+    // Check allocator assumptions
+    //
+    // The register allocator should have extended the lifetimes of all input and internal registers so that
+    // none interfere with the target.
+    noway_assert(addrReg != targetReg);
+
+    noway_assert(addrReg != loadReg);
+    noway_assert(dataReg != loadReg);
+
+    noway_assert((treeNode->OperGet() == GT_XCHG) || (addrReg != dataReg));
+
+    assert(addr->isUsedFromReg());
+    noway_assert(tempReg != REG_NA);
+    noway_assert(tempReg != targetReg);
+    noway_assert((targetReg != REG_NA) || (treeNode->OperGet() != GT_XCHG));
+
+    // Store exclusive unpredictable cases must be avoided
+    noway_assert(tempReg != addrReg);
+
+    // NOTE: `genConsumeAddress` marks the consumed register as not a GC pointer, as it assumes that the input
+    // registers
+    // die at the first instruction generated by the node. This is not the case for these atomics as the  input
+    // registers are multiply-used. As such, we need to mark the addr register as containing a GC pointer until
+    // we are finished generating the code for this node.
+
+    gcInfo.gcMarkRegPtrVal(addrReg, addr->TypeGet());
+
+    // Emit code like this:
```

Comment:
I know this is done for other platforms as well, but not sure why this cannot be done in lower. The codegen should just emit the code instead of inserting the loop like code here.

Reply:
I assume the reason is that we'd need new nodes for the atomic loads and stores to create the loop with there. I feel like changing that should be a separate PR from this if desired.